pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/LLCSegmentCompletionHandlers.java

```diff
@@ -0,0 +1,327 @@
+    }
+
+    LOGGER.info("Response: instance={}  segment={} status={} offset={}", requestParams.getInstanceId(), requestParams.getSegmentName(),
+        response.getStatus(), response.getOffset());
+
+    return response.toJsonString();
+  }
+
+  // This method may be called in any controller, leader or non-leader.
+  @POST
+  @Path(SegmentCompletionProtocol.MSG_TYPE_SEGMENT_UPLOAD)
+  @Produces(MediaType.APPLICATION_JSON)
+  @Consumes(MediaType.MULTIPART_FORM_DATA)
+  public String segmentUpload(
+      @QueryParam(SegmentCompletionProtocol.PARAM_INSTANCE_ID) String instanceId,
+      @QueryParam(SegmentCompletionProtocol.PARAM_SEGMENT_NAME) String segmentName,
+      @QueryParam(SegmentCompletionProtocol.PARAM_OFFSET) long offset,
+      FormDataMultiPart multiPart
+  ) {
+    SegmentCompletionProtocol.Request.Params requestParams = new SegmentCompletionProtocol.Request.Params();
+    requestParams.withInstanceId(instanceId).withSegmentName(segmentName).withOffset(offset);
+    LOGGER.info(requestParams.toString());
+
+    final String segmentLocation = uploadSegment(multiPart, instanceId, segmentName, true);
+    if (segmentLocation == null) {
+      return SegmentCompletionProtocol.RESP_FAILED.toJsonString();
+    }
+    SegmentCompletionProtocol.Response.Params responseParams = new SegmentCompletionProtocol.Response.Params()
+        .withOffset(requestParams.getOffset())
+        .withSegmentLocation(segmentLocation)
+        .withStatus(SegmentCompletionProtocol.ControllerResponseStatus.UPLOAD_SUCCESS);
+
+    return new SegmentCompletionProtocol.Response(responseParams).toJsonString();
+  }
+
+  private @Nullable String uploadSegment(FormDataMultiPart multiPart, final String instanceId, final String segmentName,
+      boolean isSplitCommit) {
+    Map<String, List<FormDataBodyPart>> map = multiPart.getFields();
+    if (!PinotSegmentUploadRestletResource.validateMultiPart(map, segmentName)) {
+      return null;
+    }
+    final String name = map.keySet().iterator().next();
+    final FormDataBodyPart bodyPart = map.get(name).get(0);
+    InputStream is = bodyPart.getValueAs(InputStream.class);
+    String segmentLocation;
+
+    FileOutputStream os = null;
+    try {
+      FileUploadPathProvider provider = new FileUploadPathProvider(_controllerConf);
+      File tmpFile = new File(provider.getFileUploadTmpDir(), name + "." + UUID.randomUUID().toString());
+      tmpFile.deleteOnExit();
+      os = new FileOutputStream(tmpFile);
+      IOUtils.copyLarge(is, os);
+      os.flush();
+      LLCSegmentName llcSegmentName = new LLCSegmentName(segmentName);
+      final String rawTableName = llcSegmentName.getTableName();
+      final File tableDir = new File(provider.getBaseDataDir(), rawTableName);
+      File segmentFile;
+      if (isSplitCommit) {
+        String uniqueSegmentFileName = SegmentCompletionUtils.generateSegmentFileName(segmentName);
```

Comment:
This is the code you wrote :-)

We generate the target segment file name, and move the contents of the tmp file into that destination.